KLR/Trace/Term.lean

```diff
@@ -252,31 +227,221 @@ where
       if <- termLt x y then return true
       else return (<- termEq x y) && (<- listLt xs ys)
 
-def cmpOp : CmpOp -> Term -> Term -> Trace Bool
-  | .eq, l, r => termEq l r
-  | .ne, l, r => return not (<- termEq l r)
-  | .lt, l, r => termLt l r
-  | .le, l, r => return (<- termEq l r) || (<- termLt l r)
-  | .gt, l, r => return not (<- termEq l r) && not (<- termLt l r)
-  | .ge, l, r => return not (<- termLt l r)
-  | .is, l, r => termIsIdentical l r
-  | .isNot, l, r => return not (<- termIsIdentical l r)
-  | .isIn, l, r => termIn l r
-  | .notIn, l, r => return not (<- termIn l r)
-
--- Python comparison chains are short-circuting
--- e.g. x < y < z  => x < y && y < z
-def compare : Term -> List CmpOp -> List Term -> Trace Term
-  | x, [op], [y] => return bool (<- cmpOp op x y)
-  | x, op::ops, y::ys => do
-     if (<- cmpOp op x y)
-     then compare y ops ys
-     else return (bool false)
-  | _, _, _ => throw "invalid comparison"
-where
-  bool b := .expr (.value $ .bool b) .bool
+local instance : Coe Bool Term where
+  coe b := .expr (.value $ .bool b) .bool
+
+def binop (op : BinOp) (l r : Term) : Trace Term := do
+  match op with
+  -- logical
+  | .land => return ((<- l.isTrue) && (<- r.isTrue))
+  | .lor  => return ((<- l.isTrue) || (<- r.isTrue))
```

Comment:
These aren't short circuiting. Prob doesn't matter, but the short circuit version is

  | .lor => return <- (l.isTrue <||> r.isTrue)